R/derive_extreme_event.R

```diff
@@ -0,0 +1,179 @@
+#'   1. The variables specified by the `set_values_to` parameter are added to
+#'   the selected observations.
+#'   1. The observations are added to input dataset.
+#'
+#'
+#' @return The input dataset with the best or worst observation of each by group
+#'   added as new observations.
+#'
+#' @family der_prm_bds_findings
+#' @keywords der_prm_bds_findings
+#'
+#' @export
+#'
+#' @examples
+#' library(tibble)
+#'
+#' adqs <- tribble(
+#'   ~USUBJID, ~PARAMCD,       ~AVALC,
+#'   "1",      "NO SLEEP",     "N",
+#'   "1",      "WAKE UP",      "N",
+#'   "1",      "FALL ASLEEP",  "N",
+#'   "2",      "NO SLEEP",     "N",
+#'   "2",      "WAKE UP",      "Y",
+#'   "2",      "FALL ASLEEP",  "N",
+#'   "3",      "NO SLEEP",     NA_character_
+#' )
+#'
+#' # Add a new record for each USUBJID storing the the worst sleeping problem.
+#' derive_extreme_event(
+#'   adqs,
+#'   by_vars = exprs(USUBJID),
+#'   events = list(
+#'     event(
+#'       condition = PARAMCD == "NO SLEEP" & AVALC == "Y",
+#'       set_values_to = exprs(AVALC = "No sleep", AVAL = 1)
+#'     ),
+#'     event(
+#'       condition = PARAMCD == "WAKE UP" & AVALC == "Y",
+#'       set_values_to = exprs(AVALC = "Waking up more than three times", AVAL = 2)
+#'     ),
+#'     event(
+#'       condition = PARAMCD == "FALL ASLEEP" & AVALC == "Y",
+#'       set_values_to = exprs(AVALC = "More than 30 mins to fall asleep", AVAL = 3)
+#'     ),
+#'     event(
+#'       condition = all(AVALC == "N"),
+#'       set_values_to = exprs(
+#'         AVALC = "No sleeping problems", AVAL = 4
+#'       )
+#'     ),
+#'     event(
+#'       condition = TRUE,
+#'       set_values_to = exprs(AVALC = "Missing", AVAL = 99)
+#'     )
+#'   ),
+#'   order = exprs(AVAL),
+#'   mode = "first",
+#'   set_values_to = exprs(
+#'     PARAMCD = "WSP",
+#'     PARAM = "Worst Sleeping Problems"
+#'   )
+#' )
+derive_extreme_event <- function(dataset,
+                                 by_vars = NULL,
+                                 events,
+                                 order,
+                                 mode,
+                                 check_type = "warning",
+                                 set_values_to) {
+  # Check input parameters
+  assert_vars(by_vars, optional = TRUE)
+  assert_list_of(events, "event")
+  assert_order_vars(order)
+  mode <- assert_character_scalar(mode, values = c("first", "last"), case_sensitive = FALSE)
+  check_type <-
+    assert_character_scalar(
+      check_type,
+      values = c("none", "warning", "error"),
+      case_sensitive = FALSE
+    )
+  assert_varval_list(set_values_to)
+
+  # Create new observations
+  ## Create a dataset (selected_records) from `events`
+  condition_ls <- map(events, 1)
+  set_values_to_ls <- map(events, 2)
+  selected_records_ls <- map2(
+    condition_ls, set_values_to_ls,
+    function(x, y) {
+      dataset %>%
+        group_by(!!!by_vars) %>%
+        filter(!!enexpr(x)) %>%
+        mutate(!!!y) %>%
```

Comment:
Could we add a temporary event number variable and add it to the `order` in the `filter_extreme()` call? My intention is that the events are handled like the conditions in a `case_when()`. I.e., the first event from the `events` list is selected and the `order` and `mode` arguments are only used for the case that there is more than one observation of a particular event. Then the user would not need to create an order variable for the events.
Consider for example that `AVAL` should be `0` for the `"Missing"` event. Currently `AVAL` could not be used for `order` but the user would need to create a temporary order variable and remove it after the `derive_extreme_event()` call.
Does this make sense?

Reply:
I updated the documentation as "Add the first available record from `events`...", so that 
1. for all selected events, always choose the first available event by entering order;
2. `order`/`mode` in this function is for filtering extreme record within a certain event (when there is more than one observation of a particular event)